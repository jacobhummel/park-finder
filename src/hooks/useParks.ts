import { useQuery } from "react-query";

export interface Park {
  location: number[];
  name: string;
}

interface ParksResponse {
  results: Park[];
}

const DEMO_RESULTS = [
  { name: "French Regional Park", location: [32, 45] },
  { name: "Knollwood Park", location: [32, 45] },
  { name: "Gro Tonka Park", location: [32, 45] },
];

const waitFor = (delay: number) =>
  new Promise((resolve) => setTimeout(resolve, delay));

export default function useParks() {
  return useQuery<ParksResponse>({
    queryFn: async () => {
      await waitFor(1000); // fake network speed

      return Promise.resolve({
        results: DEMO_RESULTS,
      });
    },
  });
}
