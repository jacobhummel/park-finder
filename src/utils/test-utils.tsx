import { AppCtxProvider } from "../context";
import React from "react";
import { render as testingLibRender } from "@testing-library/react-native";
import { QueryClientProvider } from "react-query";
import { NavigationContainer } from "@react-navigation/native";
import { NativeBaseProvider } from "native-base";
import client from "api/client";
import customTheme from "constants/nativeBaseTheme";

export function render(element: React.ReactElement) {
  // https://docs.nativebase.io/next/testing
  const inset = {
    frame: { x: 0, y: 0, width: 0, height: 0 },
    insets: { top: 0, left: 0, right: 0, bottom: 0 },
  };

  return testingLibRender(
    <AppCtxProvider>
      <QueryClientProvider client={client}>
        <NavigationContainer theme={customTheme}>
          <NativeBaseProvider initialWindowMetrics={inset}>
            {element}
          </NativeBaseProvider>
        </NavigationContainer>
      </QueryClientProvider>
    </AppCtxProvider>
  );
}
