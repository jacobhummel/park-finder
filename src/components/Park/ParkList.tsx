import { HStack, FlatList, Spinner, Heading } from "native-base";
import { useNavigation } from "@react-navigation/native";

import useParks from "hooks/useParks";
import ParkRow from "./ParkRow";
import { StackList } from "navigation/RootStack";
import { useCallback } from "react";

const ParkList = ({ searchText }: { searchText: string }) => {
  const { isFetching, isLoading, data, refetch } = useParks(searchText);
  const { navigate } = useNavigation<StackList>();
  const handleRowPress = useCallback(
    (id: string) => {
      navigate("MapScreen", { id });
    },
    [navigate]
  );

  if (isLoading) {
    return (
      <HStack space={2} justifyContent="center">
        <Spinner accessibilityLabel="Loading parks" />
        <Heading color="primary.500" fontSize="md">
          Loading
        </Heading>
      </HStack>
    );
  }

  return (
    <FlatList
      w="100%"
      data={data?.results || []}
      refreshing={isFetching}
      onRefresh={refetch}
      renderItem={({ item }) => <ParkRow {...item} onPress={handleRowPress} />}
      keyExtractor={(item) => item.id}
    />
  );
};

export default ParkList;
