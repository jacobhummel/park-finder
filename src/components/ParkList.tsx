import { StyleSheet } from "react-native";
import React from "react";
import useParks from "../hooks/useParks";
import {
  Box,
  HStack,
  VStack,
  Spacer,
  Text,
  FlatList,
  Spinner,
  Heading,
} from "native-base";

const ParkList = () => {
  const { isFetching, isLoading, data, refetch } = useParks();

  if (isLoading || isFetching) {
    return (
      <HStack space={2} justifyContent="center">
        <Spinner accessibilityLabel="Loading parks" />
        <Heading color="primary.500" fontSize="md">
          Loading
        </Heading>
      </HStack>
    );
  }

  return (
    <FlatList
      w="100%"
      data={data?.results || []}
      refreshing={isLoading}
      onRefresh={refetch}
      renderItem={({ item }) => (
        <Box
          borderBottomWidth="1"
          _dark={{
            borderColor: "gray.600",
          }}
          borderColor="coolGray.200"
          pl="4"
          pr="5"
          py="2"
        >
          <HStack space={3} justifyContent="space-between">
            <VStack>
              <Text
                _dark={{
                  color: "warmGray.50",
                }}
                color="coolGray.800"
                bold
              >
                {item.name}
              </Text>
              <Text
                color="coolGray.600"
                _dark={{
                  color: "warmGray.200",
                }}
              >
                {item.location}
              </Text>
            </VStack>
            <Spacer />
          </HStack>
        </Box>
      )}
      keyExtractor={(item) => item.name}
    />
  );
};

export default ParkList;

const styles = StyleSheet.create({});
