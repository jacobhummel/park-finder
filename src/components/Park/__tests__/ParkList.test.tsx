import { render } from "../../../utils/test-utils";
import ParkList, { ParkListProps } from "../ParkList";
import useParks from "../../../hooks/useParks";

// Solves TypeScript Errors
const mockedUseParks = useParks as jest.Mock<any>;

// Mock the module
jest.mock("hooks/useParks");

describe("ParkList", () => {
  beforeEach(() => {
    mockedUseParks.mockImplementation(() => ({ isLoading: true }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("should render", () => {
    // ARRANGE
    const testProps: ParkListProps = {
      searchText: "",
    };

    // ACT
    const { toJSON } = render(<ParkList {...testProps} />);

    // ASSERT
    expect(toJSON()).toMatchSnapshot();
  });

  test("should display loading indicator", () => {
    // ARRANGE
    const testProps: ParkListProps = {
      searchText: "",
    };

    // ACT
    const { getByA11yLabel } = render(<ParkList {...testProps} />);

    // ASSERT
    expect(getByA11yLabel(/loading/i)).toBeTruthy();
  });

  test("should render list with items", () => {
    // ARRANGE
    const testProps: ParkListProps = {
      searchText: "",
    };

    const MOCK_RESULTS = {
      results: [
        {
          id: "1",
          name: "Parkers Lake Park",
          address: "15205 County Rd 6 & Niagara Lane, Plymouth, MN 55447",
          lat: 44.9964,
          lng: -93.4747,
        },
        {
          id: "2",
          name: "Knollway Park",
          address: "895 Horn Dr, Minnetonka, MN 55305",
          lat: 44.9736,
          lng: -93.4525,
        },
        {
          id: "3",
          name: "Meadow Neighborhood Park",
          address: "2725 Oakland Rd, Minnetonka, MN 55305",
          lat: 44.952,
          lng: -93.4507,
        },
        {
          id: "4",
          name: "West Medicine Lake Park",
          address: "1920 W Medicine Lake Dr, Plymouth, MN 55441",
          lat: 45.0021,
          lng: -93.4351,
        },
      ],
    };
    mockedUseParks.mockImplementation(() => ({
      isLoading: false,
      data: MOCK_RESULTS,
    }));

    // ACT
    const { queryByA11yLabel, getByText } = render(<ParkList {...testProps} />);

    // ASSERT
    expect(queryByA11yLabel(/loading/i)).toBeFalsy();

    expect(getByText(MOCK_RESULTS.results[0].name)).toBeTruthy();
    expect(getByText(MOCK_RESULTS.results[0].address)).toBeTruthy();
    expect(getByText(MOCK_RESULTS.results[1].name)).toBeTruthy();
    expect(getByText(MOCK_RESULTS.results[1].address)).toBeTruthy();
    expect(getByText(MOCK_RESULTS.results[2].name)).toBeTruthy();
    expect(getByText(MOCK_RESULTS.results[2].address)).toBeTruthy();
  });
});
