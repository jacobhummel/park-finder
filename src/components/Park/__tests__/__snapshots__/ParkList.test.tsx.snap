// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParkList should render 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={Object {}}
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      accessibilityLabel="loading"
      accessible={true}
      animating={true}
      color="#0891b2"
      hidesWhenStopped={true}
      size="small"
      style={
        Array [
          Array [
            Object {},
            Object {
              "accessibilityLabel": "Loading parks",
              "dataSet": Object {},
            },
          ],
          undefined,
        ]
      }
    />
    <View
      dataSet={Object {}}
      style={
        Object {
          "width": 8,
        }
      }
    />
    <Text
      dataSet={Object {}}
      style={
        Object {
          "backgroundColor": undefined,
          "color": "#06b6d4",
          "fontFamily": undefined,
          "fontSize": 16,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 20,
          "textDecorationLine": undefined,
        }
      }
    >
      Loading
    </Text>
  </View>
</RNCSafeAreaProvider>
`;
